### DISCUSSION TITLE
[v4] how to generate catalogs from external source

### DISCUSSION THREAD

**Discussion by batusai513** [lingui/js-lingui#1614] [Q&A] [answered]
Hi, first, thank you for your work on this library, it is really appreciated.

So the thing is that the flow on the projects that I'm working goes as follows:

- add translations with explicit translations ids
- use phrase-app to add those ids with the translation
- when building the application for dev or prod, we pull the translations from phrase-app as json, and convert the json to .po using `i18next-conv`, then compile with lingui

This flow is working in v3, but testing in v4 with the new translation id hashing translations are not working anymore, after some debugging and checking the v4 changes I see that for .po files generated by lingui it is adding a `explicit-id` comment that will avoid hashing when compiling, so my question would be, is there a way to get this functionality (adding explicit-id) from your set of libraries to the above described flow?

thank you.

**Comment by andrii-bodnar** [comment]
Hi, 

There is a `printLinguiId` option in the [PO formatter configuration](https://github.com/lingui/js-lingui/tree/next/packages/format-po#usage).

By default, it has a `false` value. It allows printing `js-lingui-id: x` statement in extracted comments section

Please let us know if it helps you

**Reply by batusai513** [reply]
I'm adding the formatter to the config like this but it is getting the same result, one thing that I forgot to mention is that we are not using `lingui extract` at any point of the flow:
```js
const { formatter } = require('@lingui/format-po');
console.log('LINGUI CONFIG');
module.exports = {
  locales: [
    'de',
    'de-at',
    'en',
    'es',
    'fr-be',
    'fr',
    'it',
    'nl-be',
    'nl',
    'sv-se',
    'sv',
    'pl',
    'da',
    'pt',
  ],
  sourceLocale: 'de',
  catalogs: [
    {
      path: '<rootDir>/translations/wkda-sf/{locale}/messages',
      include: ['<rootDir>'],
      exclude: ['**/node_modules/**'],
    },
  ],
  format: formatter({ printLinguiId: true }),
};
```

**Comment by timofei-iatsenko** [comment]
> when building the application for dev or prod, we pull the translations from phrase-app as json, and convert the json to .po using i18next-conv, then compile with lingui

You can use [custom formatter](https://js-lingui-git-next-lingui.vercel.app/guides/custom-formatter) and consume json from phrase-app directly without converting it to the `po`

**Reply by batusai513** [reply]
At this moment is quite difficult to migrate from .po files because we are also adding manual overrides to translations locally via po files foe each app, and migrate all those apps will take lots of time and sync between teams, if possible I would put this as the last resource.

**Reply by timofei-iatsenko** [reply]
Just take source code of original po formatter from this repo, and adjust it as you wish. For example you could remove the part which is generating `lingui-id` from `message + context` and use provided `msgid` directly

**Reply by timofei-iatsenko** [reply, marked as answer]
```ts
import PO from "pofile"
import { CatalogFormatter, CatalogType, MessageType } from "@lingui/conf"

type POItem = InstanceType<typeof PO.Item>

const splitOrigin = (origin: string) => {
  const [file, line] = origin.split(":")
  return [file, line ? Number(line) : null] as [file: string, line: number]
}

function deserialize(items: POItem[]): CatalogType {
  return items.reduce<CatalogType<POCatalogExtra>>((catalog, item) => {
    const message: MessageType<POCatalogExtra> = {
      translation: item.msgstr[0],
      comments: item.extractedComments || [],
      context: item.msgctxt ?? null,
      obsolete: item.flags.obsolete || item.obsolete,
      origin: (item.references || []).map((ref) => splitOrigin(ref)),
    }
    
    catalog[item.msgid] = message
    return catalog
  }, {})
}

export function formatter(): CatalogFormatter {
  return {
    catalogExtension: ".po",
    templateExtension: ".pot",

    parse(content): CatalogType {
      const po = PO.parse(content)
      return deserialize(po.items)
    },

    serialize(catalog, ctx): string {
       // because you don't use extract, you don need serialize
      return "";
    },
  }
}

```

**Reply by batusai513** [reply]
Good stuff, this worked as expected, thanks for your help 🙇🏼 @thekip @andrii-bodnar
